Close #8: Update Research Dataset for Optimizing Functional Transition Metal Complexes

ZhangdeSong1 opened the pull request:
Description:

This PR addresses the issue #8 (Update research dataset: Optimizing functional transition metal complexes) by providing an updated dataset and related tasks for optimization of functional transition metal complexes (TMCs). The following updates are included:

Dataset Overview:
	•	Current Dataset Location: Initially, the dataset was stored on a private Hugging Face account under the name PeterOldriver/tmc_fewshot_initalSample.
	•	Data Details: The dataset is generated by randomly selecting 10 TMCs from a larger dataset. These TMCs feature ligands centered around a Palladium (Pd) metal in a square planar structure, with the ligands chosen from a set of 50 different options. (Detail of the Large tmc dataset can be found in [here](https://zenodo.org/records/14328055/files/ground_truth_fitness_values.csv))

Task Overview:
	•	Task Description: Given a set of initial sample TMCs and their ground truth data, along with a set of ligands that can be used to create a TMC, we ask the LLM to generate new TMCs using the provided ligands. These TMCs should be optimized for specific properties and directed towards achieving certain optimization goals.
	•	Optimization Direction: The properties can be set to “gap” (HOMO-LUMO gap) or “polar” (polarizability), with the direction of optimization set to maximize the property.

Prompt Setup:
	•	Initial TMCs: 10 TMCs are used as the initial prompt for the model.
	•	Generation: The model is asked to generate 5 new TMCs each time, based on the provided ligands and optimization properties (gap or polar).

Evaluation Method:
	•	Expected Output: The model should generate 5 TMCs in the requested format. If the response matches the expected format, the properties of the 5 generated TMCs are extracted and evaluated.
	•	The properties’ mean value will be calculated and submitted using the lm-evaluation-harness package.
       Error Handling:
	•	If the response is not in the correct format or if no TMC values can be extracted, the mean property values will be set to zero.
	•	If the response contains fewer than 5 TMCs, the mean value of the extracted properties will be adjusted by dividing the total sum by 5.
	•	If there are duplicate TMCs or TMCs outside the 50-ligand space, these entries will be discarded.

Testing:

To test this update, the following steps are required:
	1.	Install lm-eval-harness: Ensure the package is installed and switch to the current branch.
	2.	Set API keys (for commercial API models like OpenAI and Anthropic):
```
export OPENAI_API_KEY=your_openai_api_key
export ANTHROPIC_API_KEY=your_anthropic_api_key
```

3.	Run Test Commands (limited to the first 10 questions):
	•	For Anthropic API (Claude model):
```
lm_eval --model anthropic-chat-completions --model_args model=claude-3-5-sonnet-20240620 --tasks 50Ligands-TMC-gap --limit 10 --output output/anthropic-3.5 --apply_chat_template --log_samples
lm_eval --model anthropic-chat-completions --model_args model=claude-3-5-sonnet-20240620 --tasks 50Ligands-TMC-polar --limit 10 --output output/anthropic-3.5 --apply_chat_template --log_samples
```
	•	For OpenAI API (GPT-4 model):
```
lm_eval --model openai-chat-completions --model_args model=gpt-4o --tasks 50Ligands-TMC-gap --output output/gpt-4o --apply_chat_template --log_samples
lm_eval --model openai-chat-completions --model_args model=gpt-4o --tasks 50Ligands-TMC-polar --output output/gpt-4o --apply_chat_template --log_samples
```

HelloJocelynLu commented on lm_eval/tasks/science/research_topics/tmc_polar.yaml:
```diff
@@ -0,0 +1,113 @@
+  {{tmc10}}
+
+  Grounded on your chemistry knowledge, look at the pattern of the provided data and think about what makes a TMC with large polarisability.
+  Then please propose TEN *NEW* TMCs that have polarisability larger than all the TMCs above.
```
Should be "FIVE"?

HelloJocelynLu commented on lm_eval/tasks/science/research_topics/utils.py:
```diff
@@ -0,0 +1,194 @@
+from datasets import load_dataset
+import pandas as pd
+
+def process_smiles_gap(doc, results):
```
Consider removing "smiles" from the function name since it's not used in the processing code.

HelloJocelynLu commented on lm_eval/tasks/science/research_topics/utils.py:
```diff
@@ -0,0 +1,194 @@
+        "mean_homo_lumo_gap": mean_homo_lumo_gap
+    }
+
+def process_smiles_polar(doc, results):
```
Same here. Consider removing "smiles" from the function name since it's not used in the processing code.

HelloJocelynLu reviewed (commented):
1. Can we create a task group for items labeled "10_ligand_tmc_opt"? Reference documentation for group configuration is available [here](https://github.com/EleutherAI/lm-evaluation-harness/blob/main/docs/new_task_guide.md#advanced-group-configs). 
2. The property averages are difficult to compare due to different scales. Could we implement MinMax scaling to normalize all values between 0-1 for better comparability?

HelloJocelynLu commented:
Results of deepseek-r1:7b:
```bash
2025-02-18:01:50:48,302 INFO     [evaluation_tracker.py:287] Saving per-sample results for: 50Ligands-TMC-polar
local-chat-completions (model=deepseek-r1,base_url=http://localhost:11434/v1/chat/completions), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 1
|       Tasks       |Version|Filter|n-shot|      Metric       |   |Value|   |Stderr|
|-------------------|-------|------|-----:|-------------------|---|----:|---|-----:|
|50Ligands-TMC-polar|Yaml   |none  |     0|mean_polarisability|↑  |    0|±  |     0|
```
```bash
2025-02-18:02:06:01,449 INFO     [evaluation_tracker.py:287] Saving per-sample results for: 50Ligands-TMC-gap
local-chat-completions (model=deepseek-r1,base_url=http://localhost:11434/v1/chat/completions), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 1
|      Tasks      |Version|Filter|n-shot|      Metric      |   |Value |   |Stderr|
|-----------------|-------|------|-----:|------------------|---|-----:|---|-----:|
|50Ligands-TMC-gap|Yaml   |none  |     0|mean_homo_lumo_gap|↑  |0.1919|±  |0.1047|
```

ZhangdeSong1 commented:
**Bug Fix and Feature Update:**
	•	_New Evaluation Matrix_:
	•	Added a new eval matrix for the generated TMCs’ property scalers, as well as the top 3 scalers using the MinMax Scaler. The MinMax values are now extracted directly from the space database for more accurate scaling.
	•	_Task Grouping_:
	•	Combined the two tasks under a single command group for streamlined evaluation.

`lm_eval --model anthropic-chat-completions --model_args model=claude-3-5-sonnet-20240620 --tasks tmc_opt --limit 2 --output output/anthropic-3.5 --apply_chat_template --log_samples --output_path results`

	This command will invoke the claude-3.5 model and run two tasks with a limit of 2 prompts each (--limit 2). This will send 4 API calls in total.

<img width="910" alt="截屏2025-02-18 16 00 03" src="https://github.com/user-attachments/assets/b6b886e7-107d-4a98-bd24-912374ed9558" />

HelloJocelynLu commented:
It works for me.
```bash
anthropic-chat-completions (model=claude-3-5-sonnet-20240620), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 1
|        Tasks         |Version|Filter|n-shot|          Metric           |   | Value  |   |Stderr |
|----------------------|-------|------|-----:|---------------------------|---|-------:|---|------:|
|tmc_opt               |    N/A|      |      |                           |   |        |   |       |
| - 50Ligands-TMC-gap  |Yaml   |none  |     0|gap_scaler                 |↑  |  0.3574|±  | 0.0281|
|                      |       |none  |     0|mean_gap                   |↑  |  1.9258|±  | 0.1219|
|                      |       |none  |     0|top_3_gap_scaler           |↑  |  0.4879|±  | 0.0286|
| - 50Ligands-TMC-polar|Yaml   |none  |     0|mean_polarisability        |↑  |357.4903|±  |10.3198|
|                      |       |none  |     0|polarisability_scaler      |↑  |  0.6393|±  | 0.0524|
|                      |       |none  |     0|top_3_polarisability_scaler|↑  |  0.7247|±  | 0.0331|
```
